---
title: "Section 2.1 Results"
output:
  html_document:
    df_print: paged
---

# Transcriptome wide analysis of CHO cell translation initiation and elongation using Ribo-seq

## Prepare for analysis
### Load libraries & functions
```{r include=FALSE}
package_list <- c(
  "tidyverse","writexl","ggpubr","viridis"
)

lapply(package_list, require, character.only = TRUE)
```

### Create results directory
```{r include=FALSE}
results_dir <- "results/section2.1/"
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}
```

# Impact of TS on cell density
```{r}
cell_density_data <- read_delim("data/cell_density_data.txt",
  "\t",
  escape_double = FALSE, trim_ws = TRUE
)

cd_summary <- cell_density_data %>%
  group_by(Experiment, condition) %>%
  summarise(mean_density = mean(cell_density)) %>%
  group_by(Experiment)

cd_summary %>%
  filter(Experiment == "Elongation") %>%
  summarise(percent_cd_diff = 100 - (mean_density[condition == "TS"] / mean_density[condition == "NTS"]) * 100)

cd_summary %>%
  filter(Experiment == "Initiation") %>%
  summarise(percent_cd_diff = 100 - (mean_density[condition == "TS"] / mean_density[condition == "NTS"]) * 100)

table_s1 <- cell_density_data %>%
  pivot_wider(names_from= c(condition),values_from=cell_density)

fn <- paste(results_dir, "Table S1.xlsx",
  sep = ""
)

suppressMessages(if (file.exists(fn)) {
  file.remove(fn)
})

write_xlsx(list(
  Cell_density=table_s1
),
path = fn,
format_headers = TRUE
)
```

## Cell density plot
```{r}
cell_density_plot <- ggboxplot(cell_density_data, x = "condition", y = "cell_density", color = "black", add = "jitter", fill = "condition") +
  facet_wrap(~Experiment, ncol = 2) +
  theme_bw() +
  stat_compare_means(method = "t.test", label.y = 2250000) +
  labs(x = "", y = "Cells/ml", fill = "") +
  theme(legend.position = "none") +
  scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
  scale_fill_manual(values = c("#D55E00", "#56B4E9"))
ggsave(plot = cell_density_plot, filename = paste(results_dir, "Figure S1.png", sep = ""), width = 6, height = 4, device = "png", dpi = 700)
cell_density_plot
```


# Analysis of read preprocessing 
## Ribo-seq
```{r}



```


```{r}
make_count_matrix <- function(dataset){
  
  samples <- c(
  "nts_r1", "nts_r2", "nts_r3", "nts_r4",
  "ts_r1", "ts_r2", "ts_r3", "ts_r4")
  
  count_path <- "results/read_counts/"
  
  # configure to match ENA sample names 
  sample_id_file <- paste0("data/riboseq_", dataset,".txt")
  
  sample_labels <- read_delim(sample_id_file, delim = "\t", escape_double = FALSE,col_names = F, trim_ws = TRUE,progress = F) %>% 
  rename(file=X1, sample_name=X2)
  
  
  # sequenced read counts
  raw_counts_file <- paste0(count_path, "riboseq_", dataset, ".raw.counts")
  
  raw_counts <- read_delim(paste0(raw_counts_file), "|",
  escape_double = FALSE, trim_ws = TRUE ) %>%
  mutate(file=str_remove(file, paste0("data/riboseq_",dataset, "/raw_data/")), 
                         ribotype = dataset, 
                         stage = "raw") %>%
  left_join(sample_labels, by = "file") %>%
  arrange(sample_name)
  
  # post trimming counts
  trimmed_counts_file <- paste0(count_path, "riboseq_", dataset, ".trimmed.counts")
 
  riboseq_trimmed <- read_delim(trimmed_counts_file, "|", 
                                  escape_double = FALSE, trim_ws = TRUE) %>%
  mutate(file=str_remove(file, paste0("data/riboseq_",dataset, "/preproccessed_data/trimmed/")), 
                         ribotype = dataset, 
                         stage = "trimmed") %>%
  left_join(sample_labels, by = "file") %>%
  arrange(sample_name)
  

  # not aligned to rRNA
  rrna_counts_file <- paste0(count_path, "riboseq_", dataset, ".rRNA_filter.counts")
  
  riboseq_rRNA <- read_delim(paste0(rrna_counts_file),"|",escape_double = FALSE, trim_ws = TRUE) %>%
   mutate(file=str_remove(file, paste0("data/riboseq_",dataset, "/preproccessed_data/rRNA_filter/")), 
                         ribotype = dataset, 
                         stage = "rRNA_retained",
          sample_name=samples) 
  
   # not aligned to snoRNA
  snorna_counts_file <- paste0(count_path, "riboseq_", dataset, ".snoRNA_filter.counts")
  
  riboseq_snoRNA <- read_delim(paste0(snorna_counts_file),"|",escape_double = FALSE, trim_ws = TRUE) %>%
   mutate(file=str_remove(file, paste0("data/riboseq_",dataset, "/preproccessed_data/snoRNA_filter/")), 
                         ribotype = dataset, 
                         stage = "snoRNA_retained",
          sample_name=samples)
  
  # not aligned to tRNA
  trna_counts_file <- paste0(count_path, "riboseq_", dataset, ".tRNA_filter.counts")
  
  riboseq_tRNA <- read_delim(paste0(trna_counts_file),"|",escape_double = FALSE, trim_ws = TRUE) %>%
   mutate(file=str_remove(file, paste0("data/riboseq_",dataset, "/preproccessed_data/tRNA_filter/")), 
                         ribotype = dataset, 
                         stage = "tRNA_retained",
          sample_name=samples)
    
  # final retained
  final_counts_file <- paste0(count_path, "riboseq_", dataset, ".final.counts")
  riboseq_passed <- read_delim(paste0(final_counts_file),
  "|",escape_double = FALSE, trim_ws = TRUE) %>% 
    mutate(file=str_remove(file, paste0("data/riboseq_",dataset, "/preproccessed_data/complete/")), 
                         ribotype = dataset, 
                         stage = "passed",
          sample_name=samples)
   
 
   
  all_counts <- bind_rows(raw_counts, riboseq_trimmed, riboseq_rRNA, riboseq_tRNA,
  riboseq_snoRNA, riboseq_passed) %>%
  select(
    sample_name,
    ribotype, stage,
    raw_read_number
  ) %>% pivot_wider(id_cols = c(sample_name, ribotype), 
                    names_from = stage, 
                    values_from = raw_read_number)

  
preprocessing_counts <- all_counts %>%
  mutate(
    Cutadapt = raw - trimmed,
    rRNA = trimmed - rRNA_retained,
    snoRNA = rRNA_retained - snoRNA_retained,
    tRNA = snoRNA_retained - tRNA_retained,
    `<28nt or >32nt` = tRNA_retained - passed,
    Retained = passed
  ) %>%
  select(sample_name, ribotype, c("Cutadapt", "rRNA", "snoRNA", "tRNA", "<28nt or >32nt", "Retained")) %>%
  mutate(sample_name = case_when(
    sample_name == "nts_r1" ~ "NTS R1",
    sample_name == "nts_r2" ~ "NTS R2",
    sample_name == "nts_r3" ~ "NTS R3",
    sample_name == "nts_r4" ~ "NTS R4",
    sample_name == "ts_r1" ~ "TS R1",
    sample_name == "ts_r2" ~ "TS R2",
    sample_name == "ts_r3" ~ "TS R3",
    sample_name == "ts_r4" ~ "TS R4",
  ))

return(preprocessing_counts)
}
```


```{r include=FALSE}
cycloheximide_counts <- make_count_matrix("chx") 
harringtonine_counts <- make_count_matrix("harr")
nd_counts <- make_count_matrix("nd")
```


```{r}
stack_order <- c(
  "Cutadapt", "rRNA", "snoRNA", "tRNA", "<28nt or >32nt", "Retained"
)

bind_rows(cycloheximide_counts, 
         harringtonine_counts, 
         nd_counts) %>%
  mutate(
    ribotype = case_when(
      ribotype == "chx" ~ "Cycloheximide",
      ribotype == "harr" ~ "Harringtonine",
      ribotype == "nd" ~ "No drug",
    )
  )  %>%
  pivot_longer(cols = c("Cutadapt", "rRNA", "snoRNA", "tRNA", "<28nt or >32nt", "Retained"), values_to = "count", names_to = "stage") %>%
  ggplot(aes(x = sample_name, y = count, fill = factor(stage, levels = stack_order))) +
  geom_bar(stat = "identity", position = "fill") +
  theme_bw() +
  # scale_fill_brewer(palette = "Paired") +
  labs(x = "", y = "Proportion of reads", fill = "") +
  facet_wrap(~ribotype, nrow = 3) +
  scale_fill_viridis(discrete = T)

ggsave(filename = paste(results_dir, "Figure S2.png", sep = ""), width = 7, height = 8, device = "png", dpi = 700)
```
### Summary of reads per dataset
```{r}
bind_rows(cycloheximide_counts, harringtonine_counts, nd_counts) %>%
  rowwise() %>%
  mutate(raw=sum(c_across(Cutadapt:Retained))) %>% 
  group_by(ribotype) %>%
  summarise(mean(raw))

bind_rows(cycloheximide_counts, harringtonine_counts, nd_counts) %>%
  rowwise() %>%
  mutate(raw=sum(c_across(Cutadapt:Retained))) %>% 
  summarise(mean(1 - ((raw - Retained) / raw)))


bind_rows(cycloheximide_counts, harringtonine_counts, nd_counts) %>%
  group_by(ribotype) %>%
  summarise(sum(Retained))
```
## RNASeq
```{r}
rnaseq_raw <- read_delim(paste0(count_path, "rnaseq.raw.counts"),
  "|",
  escape_double = FALSE, trim_ws = TRUE
) %>%
  mutate(sample_name = samples, ribotype = "rnaseq", stage = "raw")


rnaseq_preprocessed <- read_delim(paste0(count_path, "rnaseq.preprocessed.counts"),
  "|",
  escape_double = FALSE, trim_ws = TRUE
) %>%
  mutate(sample_name = samples, ribotype = "rnaseq", stage = "pass")

rnaseq_counts <- bind_rows(
  rnaseq_raw,
  rnaseq_preprocessed
) %>%
  select(
    sample_name,
    ribotype,
    stage,
    raw_read_number
  ) %>%
  pivot_wider(id_cols = c(sample_name, ribotype), names_from = stage, values_from = raw_read_number)

rnaseq <- rnaseq_counts %>%
  mutate(
    Cutadapt = raw - pass,
    Retained = pass
  ) %>%
  select(sample_name, ribotype, c("Cutadapt", "Retained"))




rnaseq_counts %>%
  summarise(mean(raw))

rnaseq %>%
  summarise(mean(Cutadapt), mean(Retained))
```

```{r}

fn <- paste(results_dir, "Table S2.xlsx",
  sep = ""
)

suppressMessages(if (file.exists(fn)) {
  file.remove(fn)
})

write_xlsx(list(Cycloheximide = cycloheximide, Harrintonine = harringtonine, `No Drug` = nd, `RNA-seq` = rnaseq),
  path = fn,
  format_headers = TRUE
)

```


# RPF selected reads
```{r include=FALSE}
length_distribution_files <- fs::dir_ls("results/read_length_distribution/", 
                        regexp = "\\.txt$",)



length_distirbution_data <- length_distribution_files %>%
  map_dfr(read_table2, .id = "source", col_names = FALSE) %>%
  mutate(ribotype=case_when(
    str_detect(source,"riboseq_chx") ~ "riboseq_chx",
    str_detect(source,"riboseq_harr") ~ "riboseq_harr",
    str_detect(source,"riboseq_nd") ~ "riboseq_nd",
    str_detect(source,"rnaseq_se") ~ "RNA-seq")) %>%
  mutate(sample=case_when(
    str_detect(source,"nts_r1") ~ "NTS_R1",
     str_detect(source,"nts_r2") ~ "NTS_R2",
     str_detect(source,"nts_r3") ~ "NTS_R3",
     str_detect(source,"nts_r4") ~ "NTS_R4",
       str_detect(source,"ts_r1") ~ "TS_R1",
     str_detect(source,"ts_r2") ~ "TS_R2",
     str_detect(source,"ts_r3") ~ "TS_R3",
     str_detect(source,"ts_r4") ~ "TS_R4",
    str_detect(source,"merged") ~ "rnaseq")) %>%
  mutate(readlength=X2,
         count=X1) %>%
  dplyr::select(sample,ribotype,readlength, count)
```

```{r fig.height=5, fig.width=5}
labels <- c(riboseq_chx = "Ribo-seq (CHX)", riboseq_harr = "Ribo-seq (HARR)", riboseq_nd = "Ribo-seq (ND)", `RNA-seq` = "RNA-seq")

read_lengths_hist_plot <- length_distirbution_data %>%
  filter(readlength >= 25 & readlength <= 34) %>%
  mutate(ribotype = factor(ribotype, levels = c("riboseq_harr", "riboseq_chx", "riboseq_nd", "RNA-seq"))) %>%
  group_by(ribotype) %>%
  mutate(total = count) %>%
  ggplot(aes(x = readlength, y = total, fill = ifelse(readlength > 27 & readlength < 32 & ribotype != "RNA-seq", "RPF", "RNA"))) +
  geom_bar(stat = "identity") +
  facet_wrap(~ribotype, ncol = 2, labeller = labeller(ribotype = labels), scales = "free_x") +
  labs(y = "Number of reads", x = "Read length (nts)", fill = "") +
  scale_y_continuous(breaks = c(20000000, 40000000, 60000000), labels = c("20 million", "40 million", "60 million")) +
  scale_x_continuous(breaks = 25:34) +
  theme_bw() +
  theme(strip.background = element_rect(fill = "white"), legend.position = "bottom") +
  scale_fill_manual(values = c("#999999", "#009E73"))


read_lengths_hist_plot
ggsave(plot = read_lengths_hist_plot, filename = paste(results_dir, "Figure 1D.png", sep = ""), width = 5, height = 5, device = "png", dpi = 700)
```
## Phasing
```{r include=FALSE}
phasing_path <- "results/rpf_psite_analysis/"

riboseq_chx_phasing <- read_delim(paste0(phasing_path, "riboseq_chx_phasing.txt"),
  "\t",
  escape_double = FALSE, trim_ws = TRUE,
  skip = 32
) %>%
  summarise(
    `0` = mean(phase0),
    `1` = mean(phase1),
    `2` = mean(phase2)
  ) %>%
  mutate(ribotype = "Ribo-seq (CHX)")

riboseq_harr_phasing <- read_delim(paste0(phasing_path, "riboseq_harr_phasing.txt"),
  "\t",
  escape_double = FALSE, trim_ws = TRUE,
  skip = 32,
) %>%
  summarise(
    `0` = mean(phase0),
    `1` = mean(phase1),
    `2` = mean(phase2)
  ) %>%
  mutate(ribotype = "Ribo-seq (HARR)")

riboseq_nd_phasing <- read_delim(paste0(phasing_path, "riboseq_nd_phasing.txt"),
  "\t",
  escape_double = FALSE, trim_ws = TRUE,
  skip = 32
) %>%
  summarise(
    `0` = mean(phase0),
    `1` = mean(phase1),
    `2` = mean(phase2)
  ) %>%
  mutate(ribotype = "Ribo-seq (ND)")

rnaseq_phasing <- read_delim(paste0(phasing_path, "rnaseq_se_phasing.txt"),
  "\t",
  escape_double = FALSE, trim_ws = TRUE,
  skip = 32
) %>%
  summarise(
    `0` = mean(phase0),
    `1` = mean(phase1),
    `2` = mean(phase2)
  ) %>%
  mutate(ribotype = "RNA-seq")
```

## Plot the phasing figure 
```{r}
phasing_bar_plot <- bind_rows(riboseq_chx_phasing, riboseq_harr_phasing, riboseq_nd_phasing, rnaseq_phasing) %>%
  pivot_longer(cols = c(`0`, `1`, `2`), values_to = "Proportion") %>%
  mutate(ribotype = factor(ribotype, levels = c("Ribo-seq (HARR)", "Ribo-seq (CHX)", "Ribo-seq (ND)", "RNA-seq"))) %>%
  ggplot(aes(x = name, y = Proportion, fill = name)) +
  geom_bar(stat = "identity") +
  facet_wrap(~ribotype, ncol = 2) +
  labs(x = "Frame", y = "Proportion of RPFs", fill = "Frame") +
  scale_fill_viridis(discrete = T) +
  theme_bw() +
  theme(strip.background = element_rect(fill = "white"), legend.position = "bottom")

phasing_bar_plot

ggsave(plot = phasing_bar_plot, filename = paste(results_dir, "Figure 1E.png", sep = ""), width = 5, height = 5, device = "png", dpi = 700)
```
## Enrichment of RPFs in the harringtonine data at annotated TIS
Demonstration that harringtonine has a higher occupancy for annotation TIS in comparison to CHX and No-drug
```{r}


chx_metagene <- read_tsv("results/metagene_profiles/chx.metagene.txt") %>%
  mutate(ribotype = "chx")
harr_metagene <- read_tsv("results/metagene_profiles/harr.metagene.txt") %>%
  mutate(ribotype = "harr")
nd_metagene <- read_tsv("results/metagene_profiles/nd.metagene.txt") %>%
  mutate(ribotype = "nd")
metagene <- bind_rows(chx_metagene, harr_metagene, nd_metagene)

metagene <- metagene %>%
  mutate(Means = rowMeans(.[, 3:6]))

metagene <- metagene %>%
  mutate(Frame = case_when(
    position %% 3 == 0 ~ "0",
    position %% 3 == 1 ~ "1",
    position %% 3 == 2 ~ "2"
  ))

metagene$region <- factor(metagene$region, # Reordering group factor levels
  levels = c("START", "CDS", "STOP")
)

labels <- c(chx = "Cycloheximide", harr = "Harringtonine", nd = "No drug")

safe_colorblind_palette <- c("#0072B2", "#000000", "#D55E00")

start_codon_plot <- metagene %>%
  filter(region == "START") %>%
  filter(position <= 50) %>%
  mutate(ribotype = case_when(
    ribotype == "chx" ~ "Cycloheximide",
    ribotype == "harr" ~ "Harringtonine",
    ribotype == "nd" ~ "No drug"
  )) %>%
  mutate(ribotype = factor(ribotype, levels = c("Harringtonine", "Cycloheximide", "No drug"))) %>%
  ggplot(aes(x = position, y = Means, color = ribotype), alpha = ribotype) +
  geom_line(linewidth = 0.4) +
  geom_point(size = 0.8) +
  labs(y = "Averge RPF density", color = "", x = "Transcript position") +
  theme_bw() +
  theme(legend.position = "top") +
  scale_color_manual(values = safe_colorblind_palette) +
  scale_alpha_manual(values = c(0.6, 1, 0.6))
start_codon_plot
ggsave(filename = paste(results_dir, "Figure 1F.png", sep = ""), width = 6, height = 4, device = "png", dpi = 700)
```

# Knit to R script
```{r}
#knitr::purl("Results_2_1.Rmd")
```
